src/collective/mustread/tracker.py

```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+            new_users.append(userid)
+            data = dict(
+                userid=userid,
+                status=u'mustread',
+                deadline=deadline,
+                scheduled_at=now,
+                scheduled_by=current_user,
+                uid=uid,
+                type=obj.portal_type,
+                title=obj.Title(),
+                path=path,
+                site_name=hostname,
+            )
+            self._write(**data)
+        return new_users
+
+    def what_to_read(self, context=None, userid=None):
+        session = self._get_session()
+        query = session.query(MustRead.uid).filter(
+            MustRead.status != 'read').group_by(MustRead.uid)
+
+        if context is not None:
+            path = '/'.join(context.getPhysicalPath())
+            query = query.filter(MustRead.path.startswith(path))
+        if userid is not None:
+            query = query.filter(MustRead.userid == unicode(userid))
+        query = query.order_by(MustRead.path)
+        uids = [r[0] for r in self.query_all(query)]
+        for uid in uids:
+            obj = api.content.get(UID=uid)
+            if obj is None:
+                log.warning('mustread db contains broken uid: ' + uid)
+                continue
+            yield obj
+
+    def who_must_read(self, obj):
+        must_reads = self._read(uid=utils.getUID(obj), status='mustread').all()
+        return dict([(m.userid, m.deadline) for m in must_reads])
+
+    def get_report(self, context=None, include_children=True, userid=None,
+                   start_date=None):
+        path = '/'.join(api.portal.get().getPhysicalPath())
+        if context is not None:
+            path = '/'.join(context.getPhysicalPath())
+
+        session = self._get_session()
+        query = session.query(MustRead)
+
+        if start_date:
+            query = query.filter(or_(
+                MustRead.scheduled_at >= start_date.date(),
+                MustRead.read_at >= start_date.date()))
+
+        if userid:
+            query = query.filter(MustRead.userid == userid)
+        if include_children:
+            query = query.filter(MustRead.path.startswith(path))
+        else:
+            query = query.filter(MustRead.path == path)
+
+        query = query.order_by(MustRead.path)
+
+        for item in query.all():
+            # remove sqlalchemy specific data
+            # (see discussion at http://stackoverflow.com/q/1958219/810427)
+            yield dict([(k, v) for k, v in item.__dict__.iteritems()
+                        if not k.startswith('_')])
+
+    def get_stats_csv(self, csvfile, context=None, recursive=True):
+        """export mustread database entries for all objects within context
+        csvfile is a file-like object with a write method
+        (see csv.writer documentation)
+        xxx if guido does not like csv creation here in the api we can move
+        this to our contentrule
```

Comment:
Oh no this is cool. We're managing a SQL DB and this is a very valid view on that DB.